factor/telegram/classes/factor.php

```diff
@@ -0,0 +1,315 @@
+                'secret' => $newcode,
+                'label' => $_SERVER['HTTP_USER_AGENT'],
+                'timecreated' => time(),
+                'createdfromip' => $USER->lastip,
+                'timemodified' => time(),
+                'lastverified' => time(),
+                'revoked' => 0,
+            ), true);
+            $token = get_config('factor_telegram', 'telegrambottoken');
+            $telegram = new telegram($token);
+            $a = new \stdClass();
+            $a->sitename = 'Moodle'; // TODO
+            $a->code = $newcode;
+            $message = get_string('telegram:message', 'factor_telegram', $a);
+            $telegram->send_message($telegramuserid, $message);
+
+        } else if ($record->timecreated + $duration < time()) {
+            // Old code found. Keep id, update fields.
+            $DB->update_record('tool_mfa', array(
+                'id' => $record->id,
+                'secret' => $newcode,
+                'label' => $_SERVER['HTTP_USER_AGENT'],
+                'timecreated' => time(),
+                'createdfromip' => $USER->lastip,
+                'timemodified' => time(),
+                'lastverified' => time(),
+                'revoked' => 0,
+            ));
+            $instanceid = $record->id;
+            $token = get_config('factor_telegram', 'telegrambottoken');
+            $telegram = new telegram($token);
+            $a = new \stdClass();
+            $a->sitename = 'Moodle'; // TODO
+            $a->code = $newcode;
+            $message = get_string('telegram:message', 'factor_telegram', $a);
+            $telegram->send_message($telegramuserid, $message);
+        }
+    }
+
+    /**
+     * Verifies entered code against stored DB record.
+     *
+     * @return bool
+     */
+    private function check_verification_code($enteredcode) {
+        global $DB, $USER;
+        $duration = get_config('factor_telegram', 'duration');
+
+        // Get instance that isnt parent email type (label check).
+        // This check must exclude the main singleton record, with the label as the email.
+        // It must only grab the record with the user agent as the label.
+        $sql = 'SELECT *
+                  FROM {tool_mfa}
+                 WHERE userid = ?
+                   AND factor = ?
+                   AND label NOT LIKE \'telegram:%\'';
+        $record = $DB->get_record_sql($sql, array($USER->id, 'telegram'));
+
+        if ($enteredcode == $record->secret) {
+            if ($record->timecreated + $duration > time()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Cleans up email records once MFA passed.
+     *
+     * {@inheritDoc}
+     */
+    public function post_pass_state() {
+        global $DB, $USER;
+        // Delete all telegram records except base record.
+        $selectsql = 'userid = ?
+                  AND factor = ?
+                   AND label NOT LIKE \'telegram:%\'';
```

Comment:
I think there is some confusion around the intended data model and the handling of the 'secret'. You should never delete this record nor create it while logging in, just set the secret field and then unset just that one field. There should only ever be one factor record per device / account, and not new records made for each login.

I think this explains quite a few of the kinks in my testing and once fixed it should all be pretty smooth

Reply:
This is the main thing which will be affected by the planned refactor, so the post state will only remove the transient secrets and not touch the main mfa table.